pkg/segment/writer/metrics/wal/wal.go

```diff
@@ -0,0 +1,297 @@
+	}
+
+	blockSize := uint32(blockBuf.Len())
+	if err := binary.Write(w.file, binary.LittleEndian, blockSize); err != nil {
+		log.Errorf("AppendToWAL: failed to write block size to WAL file: %v", err)
+		return err
+	}
+
+	_, err = w.file.Write(blockBuf.Bytes())
+	if err != nil {
+		log.Errorf("FlushWal: failed to write block content of size %d to WAL file: %v", len(blockBuf.Bytes()), err)
+		return err
+	}
+
+	w.totalDps += uint32(len(dps))
+	w.encodedSize += uint64(blockSize)
+
+	return nil
+}
+
+func (w *WAL) Close() error {
+	return w.file.Close()
+}
+
+func DeleteWAL(baseDir, filename string) error {
+	filePath := filepath.Join(baseDir, filename)
+	return os.Remove(filePath)
+}
+
+type WalIterator struct {
+	file         *os.File
+	currentBlock []WALDatapoint
+	currentIndex int
+}
+
+func (w *WAL) GetWALStats() (string, uint32, uint64) {
+	return w.fNameWithPath, w.totalDps, w.encodedSize
+}
+
+func NewReaderWAL(baseDir, filename string) (*WalIterator, error) {
+	filePath := filepath.Join(baseDir, filename)
+	f, err := os.Open(filePath)
+	if err != nil {
+		log.Errorf("NewReaderWAL: failed to open WAL file at path %s: %v", filePath, err)
+		return nil, err
+	}
+	return &WalIterator{file: f}, nil
+}
+
+func (it *WalIterator) Next() (*WALDatapoint, bool, error) {
+	if it.currentIndex < len(it.currentBlock) {
+		dp := &it.currentBlock[it.currentIndex]
+		it.currentIndex++
+		return dp, true, nil
+	}
+
+	var blockSize uint32
+	err := binary.Read(it.file, binary.LittleEndian, &blockSize)
+	if errors.Is(err, io.EOF) {
+		log.Infof("Next: reached end of WAL file")
+		return nil, false, nil
+	} else if err != nil {
+		log.Errorf("Next: failed to read block size from WAL file: %v", err)
+		return nil, false, err
+	}
+
+	blockData := make([]byte, blockSize)
+	_, err = io.ReadFull(it.file, blockData)
+	if err != nil {
+		log.Errorf("Next: failed to read block data of size %d: %v", blockSize, err)
+		return nil, false, err
+	}
+
+	blockBuf := bytes.NewReader(blockData)
+	newBlock, err := decodeWALBlock(blockBuf)
+
+	if err != nil {
+		log.Errorf("Next: dataDecompression failed: %v", err)
+		return nil, false, err
+	}
+
+	it.currentBlock = newBlock
+	it.currentIndex = 1
+
+	return &newBlock[0], true, nil
+}
+
+func (it *WalIterator) Close() error {
+	return it.file.Close()
+}
+
+func encodeWALBlock(dps []WALDatapoint, compressedBlockBuf *bytes.Buffer) error {
+	N := uint32(len(dps))
+	if N == 0 {
+		log.Warn("EncodeWALBlock: received empty data points slice")
+		return errors.New("empty data points")
+	}
+
+	rawBlockBuf := &bytes.Buffer{}
+
+	err := binary.Write(rawBlockBuf, binary.LittleEndian, N)
+	if err != nil {
+		log.Errorf("EncodeWALBlock: failed to write datapoint count: %v", err)
+		return err
+	}
+
+	timestamps := make([]uint64, N)
+	dpVals := make([]float64, N)
+	tsids := make([]uint64, N)
+	for i, dp := range dps {
+		timestamps[i] = dp.Timestamp
+		dpVals[i] = dp.DpVal
+		tsids[i] = dp.Tsid
+	}
+
+	for _, ts := range timestamps {
```

Comment:
here you can do 
```
for _ dp := range dps {
       binary.Write(rawBlockBuf, binary.LittleEndian, dp.Timestamp)
```
I understand you are trying to save iteration of the for loops, but allocating these temp vars is a huge mem pressure.